pvlib/iam.py

```diff
@@ -527,3 +527,120 @@ def sapm(aoi, module, upper=None):
         iam = pd.Series(iam, aoi.index)
 
     return iam
+
+
+def marion_integrate(function, surface_tilt, region, N=180):
+    """
+    Integrate an incidence angle modifier (IAM) function over solid angle
+    to determine a diffuse irradiance correction factor using Marion's method.
+
+    Parameters
+    ----------
+    function : callable func(aoi)
+        The IAM function to evaluate across solid angle.  The function must
+        be vectorized and take only one parameter, the angle of incidence in
+        degrees.
+
+    surface_tilt : numeric
+        Surface tilt angles in decimal degrees.
+        The tilt angle is defined as degrees from horizontal
+        (e.g. surface facing up = 0, surface facing horizon = 90).
+
+    region : {'sky', 'horizon', 'ground'}
+        The region to integrate over.  Must be one of:
+
+            * 'sky': radiation from the sky dome
+            * 'horizon': radiation from the region of the sky near the horizon
+            * 'ground': radiation reflected from the ground
+
+        See [1]_ for a detailed description of each class.
+
+    N : int, default 180
+        The number of increments in the zenith integration.
+        If ``region=='horizon'``, a larger number should be used
+        (1800 was used in [1]_).
+
+    Returns
+    -------
+    Fd : numeric
+        AOI diffuse correction factor for the specified region.
+
+    References
+    ----------
+    .. [1] B. Marion "Numerical method for angle-of-incidence correction
+       factors for diffuse radiation incident photovoltaic modules",
+       Solar Energy, Volume 147, Pages 344-348. 2017.
+       DOI: 10.1016/j.solener.2017.03.027
+
+    Examples
+    --------
+    >>> marion_integrate(pvlib.iam.ashrae, 20, 'sky')
+    0.9596085829811408
+
+    >>> from functools import partial
+    >>> f = partial(pvlib.iam.physical, n=1.3)
+    >>> marion_integrate(f, [20, 30], 'sky')
+    array([0.96225034, 0.9653219 ])
+    """
+    beta = np.radians(surface_tilt)
+    if isinstance(beta, pd.Series):
+        # convert Series to np array for broadcasting later
+        beta = beta.values
+    ai = np.pi/N  # angular increment
+
+    phi_range = np.linspace(0, np.pi, N, endpoint=False)
+    psi_range = np.linspace(0, 2*np.pi, 2*N, endpoint=False)
+
+    # the pseudocode in [1] do these checks at the end, but it's
+    # faster to do this criteria check up front instead of later.
+    if region == 'sky':
```

Comment:
'sky' region contains 'horizon' region?

Reply:
According to Table 1 in the reference, it does indeed.  Any to-do item in this PR for that?

Reply:
My inclination is to expose the zenith angle limits that define 'sky' and 'horizon' and set defaults to Marion's values. But I don't want to complicate the interface. At a minimum, the docstring could describe what constitutes 'sky' and 'horizon'

Reply:
I'm not seeing a clean way of exposing them either.  Added zenith bounds to the docstrings for now at least.